contrib/inventory/azure_rm.py

```diff
@@ -300,24 +307,48 @@ def __init__(self, args):
         self.log("setting subscription_id")
         self.subscription_id = self.credentials['subscription_id']
 
+        # get authentication authority
+        # for adfs, user could pass in authority or not.
+        # for others, use default authority from cloud environment
+        if self.credentials.get('authority') is None:
+            self._authority = self._cloud_environment.endpoints.active_directory
+        else:
+            self._authority = self.credentials.get('authority')
+
+        # get resource from cloud environment
+        self._resource = self._cloud_environment.endpoints.active_directory_resource_id
+
         if self.credentials.get('client_id') is not None and \
            self.credentials.get('secret') is not None and \
            self.credentials.get('tenant') is not None:
             self.azure_credentials = ServicePrincipalCredentials(client_id=self.credentials['client_id'],
                                                                  secret=self.credentials['secret'],
                                                                  tenant=self.credentials['tenant'],
                                                                  cloud_environment=self._cloud_environment)
-        elif self.credentials.get('ad_user') is not None and self.credentials.get('password') is not None:
-            tenant = self.credentials.get('tenant')
-            if not tenant:
-                tenant = 'common'
-            self.azure_credentials = UserPassCredentials(self.credentials['ad_user'],
-                                                         self.credentials['password'],
-                                                         tenant=tenant,
-                                                         cloud_environment=self._cloud_environment)
+        elif self.credentials.get('ad_user') is not None and \
+             self.credentials.get('password') is not None:
+                tenant = self.credentials.get('tenant')
+                if not tenant:
+                    tenant = 'common'
+                self.azure_credentials = UserPassCredentials(self.credentials['ad_user'],
+                                                             self.credentials['password'],
+                                                             tenant=tenant,
+                                                             cloud_environment=self._cloud_environment)
```

Comment:
can't the old method be replaced entirely by adal?

Reply:
for user who're using username+password auth today, not adfs, old way works fine, but if replace that with new way, it will break because of extra credential 'client_id' is required. this is to avoid break existing users.